chisel/README.md

```diff
@@ -0,0 +1,170 @@
+
+Chisel is a fully-functional Solidity REPL, allowing you to write, execute, and debug Solidity directly in the command line.
+
+Once you finish testing, Chisel even lets you export your code to a new solidity file!
+
+In this sense, Chisel even serves as a Foundry script generator.
+
+## Feature Completion
+
+[soli](https://github.com/jpopesculian/soli) provides a great solidity REPL, achieving:
+
+- Statements
+- Custom events, errors, functions, imports
+- Inspecting a variable
+
+Chisel aims to improve upon soli, with native foundry integration by providing feature completion with:
+
+- Fork an existing chain
+- More advanced introspection
+- Better error messages and traces
+- ... many more future features!
+
+## Checklist
+
+- [ ] REPL functionality
+  - [x] Create temporary REPL contract (in memory, or temp file?).
+    - [x] Implement `forge-std/Test.sol` so that cheatcodes, etc. can be used.
+  - [x] Utilize foundry's `evm` module for REPL env.
+    - [x] Implement network forking.
+  - [ ] ~~Expression evaluation / inspection (i.e. the input `0x01 << 0x08` should inspect a `uint` of value `256`)~~
+    - Nixed due to type ambiguity. If displaying bytes is sufficient, we can still do this.
+  - [x] Input history.
+  - [ ] Use forge fmt module to format source code when printing via the `!source` command. (?)
+- [x] Cache REPL History
+  - [x] Allow a user to save/load sessions from their Chisel history.
+    - [x] Fix session loading bug wrt non-serializable `IntermediateOutput` component.
+- [ ] Custom commands / cmd flags
+  - [x] Inspect variable
+    - [x] Inspect raw stack
+  - [x] Inspect memory vars
+    - [x] Inspect raw memory
+  - [x] Inspect storage vars
+    - [ ] Inspect raw storage slots / storage layout
+  - [ ] Inspection verbosity configuration
+  - [x] Enable / disable call trace printing
+    - [x] Rip trace printing code from another module of foundry.
+  - [x] On-the-fly network forking
+  - [x] Export to file
+    - [x] Export session to script contract if within project.
+- [x] [Syntax highlighting](https://docs.rs/rustyline/10.0.0/rustyline/highlight/trait.Highlighter.html)
+- [x] Binary subcommands
+- [x] Tests.
+  - [x] Cache
+- [x] Benchmarks.
+  - [x] Session Source
+    - [x] Building
+    - [x] Executor
+    - [x] Inspection
+    - [x] Cloning
+- [ ] Optimizations.
+  - [ ] Speed up REPL execution time.
+- [x] Finish README.
+- [ ] First review.
+  - [ ] *add requested changes here*
+
+## Usage
+
+### Installation
+
+`chisel` is installed alongside Foundry cli commands!
+
+Simply run `foundryup` to install `chisel`!
+
+If you do not have `foundryup` installed, reference the Foundry [installation guide](../README.md#installation).
+
+### Cache Session
+
+While chisel sessions are not persistent by default, they can be saved to the cache via the builtin `flush` command from within the REPL.
+
+```bash
+$ chisel
+➜ uint a = 1;
+➜ uint b = a << 0x08;
+➜ !flush
```

Comment:
Would also be cool if it took an optional arg to give names to saved sessions

Reply:
Good idea- added!